cql3/Cql.g

```diff
@@ -1422,74 +1423,81 @@ idxName[cql3::index_name& name]
     | QMARK {add_recognition_error("Bind variables cannot be used for index names");}
     ;
 
-constant returns [shared_ptr<cql3::constants::literal> constant]
+constant returns [cql3::expr::untyped_constant constant]
     @init{std::string sign;}
-    : t=STRING_LITERAL { $constant = cql3::constants::literal::string(sstring{$t.text}); }
-    | t=INTEGER        { $constant = cql3::constants::literal::integer(sstring{$t.text}); }
-    | t=FLOAT          { $constant = cql3::constants::literal::floating_point(sstring{$t.text}); }
-    | t=BOOLEAN        { $constant = cql3::constants::literal::bool_(sstring{$t.text}); }
-    | t=DURATION       { $constant = cql3::constants::literal::duration(sstring{$t.text}); }
-    | t=UUID           { $constant = cql3::constants::literal::uuid(sstring{$t.text}); }
-    | t=HEXNUMBER      { $constant = cql3::constants::literal::hex(sstring{$t.text}); }
-    | { sign=""; } ('-' {sign = "-"; } )? t=(K_NAN | K_INFINITY) { $constant = cql3::constants::literal::floating_point(sstring{sign + $t.text}); }
-    ;
-
-mapLiteral returns [shared_ptr<cql3::maps::literal> map]
-    @init{std::vector<std::pair<::shared_ptr<cql3::term::raw>, ::shared_ptr<cql3::term::raw>>> m;}
+    : t=STRING_LITERAL {
+                    // This is a workaround for antlr3 not distinguishing between
+                    // calling in lexer setText() with an empty string and not calling
+                    // setText() at all.
+                    auto text = $t.text;
+                    if (text.size() == 1 && text[0] == '\xFF') {
+                        text = {};
+                    }
+                    $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::string, std::move(text)}; }
+    | t=INTEGER        { $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::integer, $t.text}; }
+    | t=FLOAT          { $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::floating_point, $t.text}; }
+    | t=BOOLEAN        { $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::boolean, $t.text}; }
+    | t=DURATION       { $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::duration, $t.text}; }
+    | t=UUID           { $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::uuid, $t.text}; }
+    | t=HEXNUMBER      { $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::hex, $t.text}; }
+    | { sign=""; } ('-' {sign = "-"; } )? t=(K_NAN | K_INFINITY) { $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::floating_point, sign + $t.text}; }
+    ;
+
+mapLiteral returns [cql3::expr::expression map]
```

Comment:
Looking at the grammar, it looks like the only use for mapLiteral is to feed its result to convert_property_map().  Can that be right?  I feel like CQL ought to allow map literals in eg, equality checks.

Reply:
there is setOrMapLiteral, which is a term.

Reply:
OK.  Then I suggest returning a collection_constructor from mapLiteral.

Reply:
Right. Changed locally, will push after more feedback.